File: scripts/python/src/tools/utils/sysinfo/bench/suites/native.py
"""Workloads measured by the bench-workloads Rust binary.

The other suites shell out to whatever the platform happens to package (7z,
sysbench, openssl), so their absolute numbers depend on how a distribution
built the tool. bench-workloads is compiled from this repository, pinned by
Cargo.lock, and dependency-free, so the same method runs on every host.
"""

import json
import os

from tools.core.paths import repo_root
from tools.core.process import capture as run
from tools.utils.sysinfo.bench.record import HIB, WORLD
from tools.utils.sysinfo.bench.suites import (
    Job,
    MeasurementError,
    Output,
    require,
    tool_path,
    version_of,
)

ITERATIONS = "800000000"
BUFFER_MIB = "256"
PASSES = "128"

RELEASE_BINARY = ("scripts", "rust", "target", "release", "bench-workloads")


def binary_path():
    override = os.environ.get("SYSINFO_BENCH_WORKLOADS")
    if override:
        return override if os.access(override, os.X_OK) else ""
    built = os.path.join(str(repo_root()), *RELEASE_BINARY)
    if os.access(built, os.X_OK):
        return built
    return tool_path("bench-workloads")


def measure(path, *args):
    result = require(run([path, *args], timeout=300), "bench-workloads")
    try:
        payload = json.loads(result.stdout)
    except json.JSONDecodeError as error:
        raise MeasurementError("bench-workloads produced unreadable output") from error
    value = payload.get("value")
    if not isinstance(value, (int, float)) or value <= 0:
        raise MeasurementError("bench-workloads reported no value")
    return float(value)


def jobs(setting):
    path = binary_path()
    if not path:
        return []
    version = version_of(path, args=("--version",))
    return [
        Job(
            name="cpu.native",
            tool="bench-workloads",
            version=version,
            method="cpu.native/1.0.0",
            outputs=(
                Output("cpu.native_single", "Mops/s", HIB, WORLD),
                Output("cpu.native_multi", "Mops/s", HIB, WORLD),
            ),
            measure=lambda: {
                "cpu.native_single": measure(
                    path, "cpu", "--threads", "1", "--iterations", ITERATIONS
                ),
                "cpu.native_multi": measure(
                    path, "cpu", "--threads", "0", "--iterations", ITERATIONS
                ),
            },
            detail={"iterations": int(ITERATIONS)},
        ),
        Job(
            name="mem.native",
            tool="bench-workloads",
            version=version,
            method="mem.native/1.0.0",
            outputs=(
                Output("mem.native_read", "GiB/s", HIB, WORLD),
                Output("mem.native_write", "GiB/s", HIB, WORLD),
            ),
            measure=lambda: {
                "mem.native_read": measure(
                    path, "memory", "--op", "read", "--mib", BUFFER_MIB, "--passes", PASSES
                ),
                "mem.native_write": measure(
                    path, "memory", "--op", "write", "--mib", BUFFER_MIB, "--passes", PASSES
                ),
            },
            detail={"buffer_mib": int(BUFFER_MIB), "passes": int(PASSES), "threads": 1},
        ),
    ]

File: scripts/rust/crates/hwtune/src/bench/suites/native.rs
use super::{Job, Measurement, job, output, require, version};
use crate::bench::runner::Setting;
use serde_json::{Value, json};
use std::{
    path::{Path, PathBuf},
    process::Command,
};
pub fn native_path() -> Result<Option<PathBuf>, String> {
    workstation::native::Resolver::discover(sysinfo::inventory::repo_root())
        .resolve("bench-workloads")
}
pub fn value_of(text: &str) -> Result<f64, String> {
    let payload: Value =
        serde_json::from_str(text).map_err(|_| "bench-workloads produced unreadable output")?;
    payload["value"]
        .as_f64()
        .filter(|n| n.is_finite() && *n > 0.0)
        .ok_or_else(|| "bench-workloads reported no value".into())
}
pub fn measure(binary: &Path, args: &[&str], seconds: u64) -> Result<f64, String> {
    value_of(&require(Command::new(binary).args(args), seconds)?)
}
pub fn jobs(setting: &Setting) -> Result<Vec<Job>, String> {
    if !setting.accepts("cpu") && !setting.accepts("mem") {
        return Ok(Vec::new());
    }
    let Some(path) = native_path()? else {
        return Ok(Vec::new());
    };
    let ver = version(&path, &["--version"], r"(\d[\d.]*)");
    let mut jobs = Vec::new();
    for (name, scale, proportion, detail, measurements) in [
        (
            "cpu.native",
            "Mops/s",
            "HIB",
            json!({"iterations":800000000}),
            vec![
                (
                    "cpu.native_single",
                    vec!["cpu", "--threads", "1", "--iterations", "800000000"],
                ),
                (
                    "cpu.native_multi",
                    vec!["cpu", "--threads", "0", "--iterations", "800000000"],
                ),
            ],
        ),
        (
            "mem.native",
            "GiB/s",
            "HIB",
            json!({"buffer_mib":256,"passes":128,"threads":1}),
            vec![
                (
                    "mem.native_read",
                    vec!["memory", "--op", "read", "--mib", "256", "--passes", "128"],
                ),
                (
                    "mem.native_write",
                    vec!["memory", "--op", "write", "--mib", "256", "--passes", "128"],
                ),
            ],
        ),
        (
            "mem.latency",
            "ns",
            "LIB",
            json!({"buffer_mib":256,"passes":4,"threads":1,"access":"dependent pointer chase"}),
            vec![(
                "mem.latency",
                vec!["memory", "--op", "latency", "--mib", "256", "--passes", "4"],
            )],
        ),
    ] {
        let binary = path.clone();
        let outputs = measurements
            .iter()
            .map(|(key, _)| output(key, scale, proportion, "world"))
            .collect();
        jobs.push(job(
            name,
            "bench-workloads",
            &ver,
            &format!("{name}/1.0.0"),
            outputs,
            detail,
            move || {
                let mut values = Vec::new();
                for (key, args) in &measurements {
                    values.push((key.to_string(), vec![measure(&binary, args, 300)?]));
                }
                Ok(Measurement::values(values))
            },
        ));
    }
    Ok(jobs)
}

#[cfg(test)]
#[path = "../../../tests/unit/bench/native_tests.rs"]
mod tests;
